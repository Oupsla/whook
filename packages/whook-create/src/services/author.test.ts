import initAuthor from './author';

describe('initAuthor', () => {
  const exec = jest.fn() as any;
  const lock = {
    take: jest.fn() as jest.Mock<undefined>,
    release: jest.fn() as jest.Mock<undefined>,
  };
  const inquirer = { prompt: jest.fn() };
  const log = jest.fn();

  beforeEach(() => {
    exec.mockReset();
    lock.take.mockReset();
    inquirer.prompt.mockReset();
    lock.release.mockReset();
    log.mockReset();
  });

  it('should work', async () => {
    exec.mockImplementationOnce((_, cb) => cb(null, 'Wayne Campbell'));
    exec.mockImplementationOnce((_, cb) => cb(null, 'wayne@warner.com'));
    lock.take.mockResolvedValueOnce(undefined);
    inquirer.prompt.mockResolvedValueOnce({
      authorName: 'Wayne Campbell',
      authorEmail: 'wayne@warner.com',
    });
    lock.release.mockResolvedValueOnce(undefined);

    const author = await initAuthor({
      inquirer,
      exec,
      lock,
      log,
    });

    expect({
      author,
      inquirerPromptCalls: inquirer.prompt.mock.calls,
      lockTakeCalls: lock.take.mock.calls,
      lockReleaseCalls: lock.release.mock.calls,
      execCalls: exec.mock.calls,
      logCalls: log.mock.calls,
    }).toMatchSnapshot();
  });

  it('should handle git failures', async () => {
    exec.mockImplementationOnce((_, cb) => cb(new Error('E_GIT_ERROR')));
    exec.mockImplementationOnce((_, cb) => cb(new Error('E_GIT_ERROR')));
    lock.take.mockResolvedValueOnce(undefined);
    inquirer.prompt.mockResolvedValueOnce({
      authorName: 'Wayne Campbell',
      authorEmail: 'wayne@warner.com',
    });
    lock.release.mockResolvedValueOnce(undefined);

    const author = await initAuthor({
      inquirer,
      exec,
      lock,
      log,
    });

    expect({
      author,
      inquirerPromptCalls: inquirer.prompt.mock.calls,
      lockTakeCalls: lock.take.mock.calls,
      lockReleaseCalls: lock.release.mock.calls,
      execCalls: exec.mock.calls,
      logCalls: log.mock.calls.filter(([type]) => 'debug-stack' !== type),
    }).toMatchSnapshot();
  });

  it('should release the lock on failure', async () => {
    exec.mockImplementationOnce((_, cb) => cb(null, 'Wayne Campbell'));
    exec.mockImplementationOnce((_, cb) => cb(null, 'wayne@warner.com'));
    lock.take.mockResolvedValueOnce(undefined);
    inquirer.prompt.mockRejectedValueOnce(new Error('E_PROMPT_ERROR'));
    lock.release.mockResolvedValueOnce(undefined);

    try {
      await initAuthor({
        inquirer,
        exec,
        lock,
        log,
      });
    } catch (err) {
      expect({
        errorCode: err.code,
        errorParams: err.params,
        inquirerPromptCalls: inquirer.prompt.mock.calls,
        lockTakeCalls: lock.take.mock.calls,
        lockReleaseCalls: lock.release.mock.calls,
        execCalls: exec.mock.calls,
        logCalls: log.mock.calls.filter(
          ([type]) => !['stack', 'debug-stack'].includes(type),
        ),
      }).toMatchSnapshot();
    }
  });
});
