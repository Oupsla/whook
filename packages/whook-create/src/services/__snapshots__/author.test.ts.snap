// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initAuthor should handle git failures 1`] = `
Object {
  "author": Object {
    "email": "wayne@warner.com",
    "name": "Wayne Campbell",
  },
  "execCalls": Array [
    Array [
      "git config --get user.name",
      [Function],
    ],
    Array [
      "git config --get user.email",
      [Function],
    ],
  ],
  "inquirerPromptCalls": Array [
    Array [
      Array [
        Object {
          "default": undefined,
          "message": "What's your name?",
          "name": "authorName",
        },
        Object {
          "default": undefined,
          "message": "You email?",
          "name": "authorEmail",
        },
      ],
    ],
  ],
  "lockReleaseCalls": Array [
    Array [
      "cli:input",
    ],
  ],
  "lockTakeCalls": Array [
    Array [
      "cli:input",
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "✍️ - Initializing author...",
    ],
    Array [
      "stack",
      undefined,
    ],
    Array [
      "stack",
      undefined,
    ],
    Array [
      "debug",
      "Could not get author from Git",
    ],
  ],
}
`;

exports[`initAuthor should release the lock on failure 1`] = `
Object {
  "errorCode": "E_PROMPT_ERROR",
  "errorParams": Array [],
  "execCalls": Array [
    Array [
      "git config --get user.name",
      [Function],
    ],
    Array [
      "git config --get user.email",
      [Function],
    ],
  ],
  "inquirerPromptCalls": Array [
    Array [
      Array [
        Object {
          "default": "Wayne Campbell",
          "message": "What's your name?",
          "name": "authorName",
        },
        Object {
          "default": "wayne@warner.com",
          "message": "You email?",
          "name": "authorEmail",
        },
      ],
    ],
  ],
  "lockReleaseCalls": Array [
    Array [
      "cli:input",
    ],
  ],
  "lockTakeCalls": Array [
    Array [
      "cli:input",
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "✍️ - Initializing author...",
    ],
  ],
}
`;

exports[`initAuthor should work 1`] = `
Object {
  "author": Object {
    "email": "wayne@warner.com",
    "name": "Wayne Campbell",
  },
  "execCalls": Array [
    Array [
      "git config --get user.name",
      [Function],
    ],
    Array [
      "git config --get user.email",
      [Function],
    ],
  ],
  "inquirerPromptCalls": Array [
    Array [
      Array [
        Object {
          "default": "Wayne Campbell",
          "message": "What's your name?",
          "name": "authorName",
        },
        Object {
          "default": "wayne@warner.com",
          "message": "You email?",
          "name": "authorEmail",
        },
      ],
    ],
  ],
  "lockReleaseCalls": Array [
    Array [
      "cli:input",
    ],
  ],
  "lockTakeCalls": Array [
    Array [
      "cli:input",
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "✍️ - Initializing author...",
    ],
  ],
}
`;
